spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/inbound/HazelcastEventDrivenMessageProducer.java

```diff
@@ -0,0 +1,147 @@
+ * @since 1.0.0
+ *
+ */
+public class HazelcastEventDrivenMessageProducer extends AbstractHazelcastMessageProducer {
+
+	public HazelcastEventDrivenMessageProducer(DistributedObject distributedObject) {
+		super(distributedObject);
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+		HazelcastIntegrationDefinitionValidator.validateCacheTypeForEventDrivenMessageProducer(getDistributedObject());
+	}
+
+	@Override
+	protected void doStart() {
+		if(getDistributedObject() instanceof IMap) {
+			setHazelcastRegisteredEventListenerId(((IMap<?, ?>)getDistributedObject()).addEntryListener(new HazelcastEntryListener(), true));
+		}
+		else if(getDistributedObject() instanceof MultiMap) {
+			setHazelcastRegisteredEventListenerId(((MultiMap<?, ?>)getDistributedObject()).addEntryListener(new HazelcastEntryListener(), true));
+		}
+		else if(getDistributedObject() instanceof ReplicatedMap) {
+			setHazelcastRegisteredEventListenerId(((ReplicatedMap<?, ?>)getDistributedObject()).addEntryListener(new HazelcastEntryListener()));
+		}
+		else if(getDistributedObject() instanceof IList) {
+			setHazelcastRegisteredEventListenerId(((IList<?>)getDistributedObject()).addItemListener(new HazelcastItemListener(), true));
+		}
+		else if(getDistributedObject() instanceof ISet) {
+			setHazelcastRegisteredEventListenerId(((ISet<?>)getDistributedObject()).addItemListener(new HazelcastItemListener(), true));
+		}
+		else if(getDistributedObject() instanceof IQueue) {
+			setHazelcastRegisteredEventListenerId(((IQueue<?>)getDistributedObject()).addItemListener(new HazelcastItemListener(), true));
+		}
+		else if(getDistributedObject() instanceof ITopic) {
+			setHazelcastRegisteredEventListenerId(((ITopic<?>)getDistributedObject()).addMessageListener(new HazelcastMessageListener()));
+		}
+	}
+
+	@Override
+	protected void doStop() {
+		if(getDistributedObject() instanceof IMap) {
+			((IMap<?, ?>)getDistributedObject()).removeEntryListener(getHazelcastRegisteredEventListenerId());
+		}
+		else if(getDistributedObject() instanceof MultiMap) {
+			((MultiMap<?, ?>)getDistributedObject()).removeEntryListener(getHazelcastRegisteredEventListenerId());
+		}
+		else if(getDistributedObject() instanceof ReplicatedMap) {
+			((ReplicatedMap<?, ?>)getDistributedObject()).removeEntryListener(getHazelcastRegisteredEventListenerId());
+		}
+		else if(getDistributedObject() instanceof IList) {
+			((IList<?>)getDistributedObject()).removeItemListener(getHazelcastRegisteredEventListenerId());
+		}
+		else if(getDistributedObject() instanceof ISet) {
+			((ISet<?>)getDistributedObject()).removeItemListener(getHazelcastRegisteredEventListenerId());
+		}
+		else if(getDistributedObject() instanceof IQueue) {
+			((IQueue<?>)getDistributedObject()).removeItemListener(getHazelcastRegisteredEventListenerId());
+		}
+		else if(getDistributedObject() instanceof ITopic) {
+			((ITopic<?>)getDistributedObject()).removeMessageListener(getHazelcastRegisteredEventListenerId());
+		}
```

Comment:
Would be some `StopCallback` abstraction better?
```
else if(getDistributedObject() instanceof ITopic) {
final ITopic<?> topic = (ITopic<?>)getDistributedObject();
final String listenerId = topic.addMessageListener(new HazelcastMessageListener()));
setStopCallback(new Runnable() {
    
    public void run() {
       topic.removeMessageListener(listenerId);
    }
});
}
```
And from here just this:
```
getStopCallback().run();
```

However we can live with your current solution, too.
Although there is need to take care about generics anyway.